### ISSUE TITLE
`field_validator('*')` not called for extra fields

### ISSUE THREAD

**Issue by sneakers-the-rat** [pydantic/pydantic#10081]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I am attempting to do some coercion for all fields, including any extra fields, but i can't do so with a `field_validator`. 

I see [the docs](https://docs.pydantic.dev/2.7/concepts/models/#extra-fields) say "by default, no validation will be applied to these extra items," but i would expect that i should be able to supply my own validation for those extra columns above and beyond the `__pydantic_extra__` annotation.

I am able to work around the problem using `@model_validator` , but it seems like all supplied fields, including extras, should be passed to `field_validator('*')`

If this is a bug rather than a wontfix, i would be happy to PR the behavior with a little guidance on what an acceptable implementation would look like since i know y'all have a high volume of issues <3

also sorry if this is a duplicate, i searched but couldn't find a previous issue

### Example Code

```Python
from typing import Any
from pydantic import BaseModel, ConfigDict, field_validator, ValidationInfo, ValidatorFunctionWrapHandler
from numpydantic import NDArray

class MyModel(BaseModel):
    model_config = ConfigDict(extra="allow")
    __pydantic_extra__: dict[str, NDArray]
    expected_field: NDArray
    
    @field_validator("*", mode="wrap")
    @classmethod
    def cast_fields_wrap(
        cls, val: Any, handler: ValidatorFunctionWrapHandler, info: ValidationInfo
    ) -> Any:
        print(f"wrap: {info.field_name}")
        return handler(val)
    
    @field_validator("*", mode="before")
    @classmethod
    def cast_fields_before(
        cls, val: Any, info: ValidationInfo
    ) -> Any:
        print(f"before: {info.field_name}")
        return val
    
    @field_validator("*", mode="after")
    @classmethod
    def cast_fields_after(
        cls, val: Any, info: ValidationInfo
    ) -> Any:
        print(f"after: {info.field_name}")
        return val

# validator not called for extra_field
instance = MyModel(expected_field = [1,2,3], extra_field = [4,5,6])
# before: expected_field
# wrap: expected_field
# after: expected_field

# but the coercion that happens from the `NDArray` type is applied
instance
# MyModel(expected_field=array([1, 2, 3]), extra_field=array([4, 5, 6]))
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.8.2
        pydantic-core version: 2.20.1
          pydantic-core build: profile=release pgo=false
                 install path: .venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.3 (main, Jun  6 2023, 17:36:45) [Clang 12.0.5 (clang-1205.0.22.9)]
                     platform: macOS-14.3.1-x86_64-i386-64bit
             related packages: pydantic-settings-2.3.4 typing_extensions-4.12.2
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
Hi!

While we don't currently support `*` validation support via field validators, you can use annotations to express before / after / wrap validation logic for extra values.

For example:

```py
from typing import Any
from typing_extensions import Annotated
from pydantic import (
    BaseModel,
    ConfigDict,
    ValidationInfo,
    ValidatorFunctionWrapHandler,
    BeforeValidator,
    WrapValidator,
    AfterValidator,
)


def cast_fields_wrap(val: Any, handler: ValidatorFunctionWrapHandler, info: ValidationInfo) -> Any:
    print(f'wrap: {info.field_name}')
    return handler(val)


def cast_fields_before(val: Any, info: ValidationInfo) -> Any:
    print(f'before: {info.field_name}')
    return val


def cast_fields_after(val: Any, info: ValidationInfo) -> Any:
    print(f'after: {info.field_name}')
    return val


MyInt = Annotated[
    int, BeforeValidator(cast_fields_before), WrapValidator(cast_fields_wrap), AfterValidator(cast_fields_after)
]


class MyModel(BaseModel):
    model_config = ConfigDict(extra='allow')

    expected_field: MyInt

    __pydantic_extra__: dict[str, MyInt]


# validation called for extra fields and your field
instance = MyModel(expected_field='1', extra_field='2')
print(repr(instance))
# MyModel(expected_field=1, extra_field=2)
```

Going to close this as not planned for now. Let me know if you have any other questions!

**Comment by sneakers-the-rat** [comment]
Thanks for the info!!!!!
